### PR TITLE
Minimal `DataViewerApp` part 2: `MainPlotWidget`

### PR DESCRIPTION
**Pull request by kangz12345**
See #144 for details.

`MainPlotWidget` shows the data with a plot (1D) or an image (2D).
It provides a signal for each data point mouse click events.

You can test it using the following script:
```python
from PyQt5.QtWidgets import QApplication
import iquip.apps.dataviewer as dv
import numpy as np

data_1d = np.random.random(10)
axes_1d = (dv.AxisInfo("Time", np.linspace(0, 9e-6, 10), "s"),)

data_2d = np.random.random((10, 10))
axes_2d = (
    dv.AxisInfo("Y", np.linspace(3e-3, 5e-3, 10), "m"),
    dv.AxisInfo("X", np.linspace(5e-3, 6e-3, 10), "m"),
)

qapp = QApplication([])
mainplots = []
for data, axes in ((data_1d, axes_1d), (data_2d, axes_2d)):
    mainplot = dv.MainPlotWidget()
    mainplots.append(mainplot)
    mainplot.setData(data, axes)
    mainplot.dataClicked.connect(lambda x: print(x))
    mainplot.show()
qapp.exec_()
```

### CODE REVIEW

**Comment by kangz12345** [comment]
I think some commits went wrong... I fixed them but please review the code thoroughly!

File: iquip/apps/dataviewer.py [hunk 1]
```diff
@@ -73,12 +77,25 @@ def setData(self, data: np.ndarray, axes: Sequence[AxisInfo]):
             if size != len(info.values):
                 raise ValueError(f"Size mismatch in {info}: expected {size} values")
 
+    def nearestDataPoint(
+        self, scenePos: pg.Point, tolerance: Optional[float] = None,  # pylint: disable=unused-argument
```
Comment (BECATRUE) [inline iquip/apps/dataviewer.py]: Please remove the trailing comma.

File: iquip/apps/dataviewer.py [hunk 2]
```diff
@@ -165,6 +197,22 @@ def setData(self, data: np.ndarray, axes: Sequence[AxisInfo]):
...
+        scenePos: pg.Point,
+        tolerance: Optional[float] = None,  # pylint: disable=unused-argument
+    ) -> Optional[Tuple[int, int]]:
+        """Overridden.
+        
+        ImageViewer does not use tolerance since it has a clear bounding box for
+          each data point.
+        """
+        dataPos = self.image.mapFromDevice(scenePos)
+        x, y = np.floor(dataPos.x()), np.floor(dataPos.y())
```
Comment (BECATRUE) [inline iquip/apps/dataviewer.py:215]: Why did you use `np.floor()`, not `int()`?

File: iquip/apps/dataviewer.py [hunk 3]
```diff
@@ -436,3 +484,77 @@ def _idToggledSlot(self, id_: int, checked: bool):
...
+        }
+        self.stack = QStackedWidget(self)
+        for plotType in MainPlotWidget.PlotType:
+            self.stack.addWidget(self.viewers[plotType].widget)
+        layout = QHBoxLayout(self)
+        layout.addWidget(self.stack)
+        # signal connection
+        for viewer in self.viewers.values():
+            viewer.plotItem.scene().sigMouseClicked.connect(
+                functools.partial(self._mouseClicked, viewer))
```
Comment (BECATRUE) [inline iquip/apps/dataviewer.py]: The line breaking looks little strange. How about this?
```python
viewer.plotItem.scene().sigMouseClicked.connect(
    functools.partial(self._mouseClicked, viewer)
)
```

File: iquip/apps/dataviewer.py [hunk 4]
```diff
@@ -73,12 +77,25 @@ def setData(self, data: np.ndarray, axes: Sequence[AxisInfo]):
...
                 raise ValueError(f"Size mismatch in {info}: expected {size} values")
 
+    def nearestDataPoint(
+        self, scenePos: pg.Point, tolerance: Optional[float] = None,  # pylint: disable=unused-argument
+    ) -> Optional[Tuple[int, ...]]:
+        """Returns the index of the nearest data point.
+        
+        Args:
+            scenePos: Scene position coordinates.
+            tolerance: Maximum Euclidean distance in the scene coordinate,
```
Comment (BECATRUE) [inline iquip/apps/dataviewer.py:89]: This is just a question! Is the unit of `tolerance` a pixel?

**Review by BECATRUE — CHANGES_REQUESTED** [review]

File: iquip/apps/dataviewer.py (hunk 1 above)
Comment (kangz12345) [reply iquip/apps/dataviewer.py]: As I said before, trailing commas help reducing diffs in the future.
For example, if I want to add more arguments to `nearestDataPoint()`, I can just add another line for the new arguments, leaving the original line.
If I didn't use that trailing comma, I would have to put a comma at the end, so the line would be modified, making a diff on that line.

In fact, the most ideal code would be:
```python
# pylint: disable=unused-argument
def nearestDataPoint(
    self,
    scenePos: pg.Point,
    tolerance: Optional[float] = None,
) -> ...:
```

File: iquip/apps/dataviewer.py (hunk 2 above)
Comment (kangz12345) [reply iquip/apps/dataviewer.py:215]: I also used it at the first time, but noticed we forgot that `int()` truncates towards zero. That means -0.5 is mapped to 0, so they are misinterpreted as valid clicks.

File: iquip/apps/dataviewer.py (hunk 4 above)
Comment (kangz12345) [reply iquip/apps/dataviewer.py:89]: I think so..! "Scene position" seems to be in pixel.

File: iquip/apps/dataviewer.py (hunk 1 above)
Comment (BECATRUE) [reply iquip/apps/dataviewer.py]: Oh, I see. It looks more efficient! Then, please never mind😃

File: iquip/apps/dataviewer.py (hunk 2 above)
Comment (BECATRUE) [reply iquip/apps/dataviewer.py:215]: Wow, I forgot it perfectly! Thanks for noticing me.

**Review by BECATRUE — APPROVED** [review]
LGTM!

**Pull request merged**